### PR TITLE
Optimize aligners over FASTQ fragments on-demand

### PR DESCRIPTION
**Pull request by smondet**
This change is [Review on Reviewable](https://reviewable.io/reviews/hammerlab/biokepi/162)

### CODE REVIEW

**Comment by smondet** [comment]
It fixes https://github.com/hammerlab/biokepi/issues/157

File: src/lib/pipeline.ml [hunk 1]
```diff
@@ -545,6 +549,67 @@ module Compiler = struct
...
+        | None -> `Not_caught
+        end
+      | Paired_end_sample (name,
+                           Gunzip_concat r1_list,
+                           Gunzip_concat r2_list)
+        when List.length r1_list = 1 -> `Not_caught
+      | other ->
+        begin match option with
+        | Some `Silent
+        | None -> `Not_caught
```
Comment (ihodes) [inline src/lib/pipeline.ml]: I think the default of when a compiler optimization cannot be made, but is expected to by the user, should be noisy alert, not silent failure and continuing with the compilation.

Comment (ihodes) [inline src/lib/pipeline.ml]: Ah I see; this is when there is no optimization specified. I recall my previous comment.

File: src/lib/pipeline.ml [hunk 2]
```diff
@@ -545,6 +549,67 @@ module Compiler = struct
           ~name:(sprintf "single-end %s"
                    (Filename.basename r1#product#path))
     in
+    let parallelize_alignment ~make_aligner sample =
+      match sample with
+      | Paired_end_sample (name,
+                           Gunzip_concat r1_list,
+                           Gunzip_concat r2_list) ->
+        let exploded =
+          let count = ref 0 in
```
Comment (ihodes) [inline src/lib/pipeline.ml]: ლ(ಥ Д ಥ )ლ

**Comment by ihodes** [comment]
LGTM!

**Pull request merged**